From e128dda22d28af6f8fd5d88f0a3a789d0b5905c1 Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Wed, 6 Aug 2014 03:48:30 -0400
Subject: [vfio] pci: Use pci_enable_msi_range() and pci_enable_msix_range()

Message-id: <20140806034829.1438.67426.stgit@gir.stowe>
Patchwork-id: 87450
O-Subject: [RHEL7.1 PATCH 41/69] vfio: Use pci_enable_msi_range() and pci_enable_msix_range()
Bugzilla: 1110896
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Dean Nelson <dnelson@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1110896
Upstream Status: 94cccde648d32abe61e2d0d1b18178f3027f1a78

commit 94cccde648d32abe61e2d0d1b18178f3027f1a78
Author: Alexander Gordeev <agordeev@redhat.com>
Date:   Fri Jan 17 17:02:21 2014 +0100

    vfio: Use pci_enable_msi_range() and pci_enable_msix_range()

    pci_enable_msix() and pci_enable_msi_block() have been deprecated; use
    pci_enable_msix_range() and pci_enable_msi_range() instead.

    [bhelgaas: changelog]
    Signed-off-by: Alexander Gordeev <agordeev@redhat.com>
    Signed-off-by: Bjorn Helgaas <bhelgaas@google.com>
    Acked-by: Alex Williamson <alex.williamson@redhat.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/vfio/pci/vfio_pci_intrs.c b/drivers/vfio/pci/vfio_pci_intrs.c
index 2103576..9dd49c9 100644
--- a/drivers/vfio/pci/vfio_pci_intrs.c
+++ b/drivers/vfio/pci/vfio_pci_intrs.c
@@ -482,15 +482,19 @@ static int vfio_msi_enable(struct vfio_pci_device *vdev, int nvec, bool msix)
   for (i = 0; i < nvec; i++)
    vdev->msix[i].entry = i;
 
-  ret = pci_enable_msix(pdev, vdev->msix, nvec);
-  if (ret) {
+  ret = pci_enable_msix_range(pdev, vdev->msix, 1, nvec);
+  if (ret < nvec) {
+   if (ret > 0)
+    pci_disable_msix(pdev);
    kfree(vdev->msix);
    kfree(vdev->ctx);
    return ret;
   }
  } else {
-  ret = pci_enable_msi_block(pdev, nvec);
-  if (ret) {
+  ret = pci_enable_msi_range(pdev, 1, nvec);
+  if (ret < nvec) {
+   if (ret > 0)
+    pci_disable_msi(pdev);
    kfree(vdev->ctx);
    return ret;
   }
-- 
1.7.1